aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp

```diff
@@ -1202,5 +1206,149 @@ Tensor _sparse_csr_prod_cpu(const Tensor& input, IntArrayRef dims_to_reduce, boo
   return result;
 }
 
+std::tuple<Tensor, Tensor> _spmm_reduce_sparse_csr_cpu(
+    const Tensor& input,
+    const Tensor& weight,
+    const c10::string_view reduce,
+    const c10::optional<Tensor>& row_indices_opt,
+    const c10::optional<Tensor>& ccol_indices_opt,
+    const c10::optional<Tensor>& csr2csc_opt) {
+
+  // See [Note: hacky wrapper removal for optional tensor]
+  c10::MaybeOwned<Tensor> row_indices_maybe_owned = at::borrow_from_optional_tensor(row_indices_opt);
+  const Tensor& row_indices = *row_indices_maybe_owned;
+  const Tensor& ccol_indices = c10::value_or_else(ccol_indices_opt, [] {return Tensor();});
+  const Tensor& csr2csc = c10::value_or_else(csr2csc_opt, [] {return Tensor();});
+
+  sparse::impl::check_spmm_reduce_inputs</*train*/false>(
+      input, Tensor(), weight, row_indices, ccol_indices, csr2csc);
+
+  auto op = sparse::impl::get_operator_enum(reduce);
+
+  auto crow = input.crow_indices();
+  auto col = input.col_indices();
+  auto val = input.values();
+
+  // init output to be all zeros, for `rows` that has no nonzero elements,
+  // the corresponding rows in the output will be zero.
+  auto out = at::zeros({input.size(0), weight.size(1)}, weight.options());
+  auto arg_out = at::empty({0}, col.options());
+
+  int64_t nnz = input._nnz();
+  if (nnz == 0) {
+    return std::make_tuple(out, arg_out);
+  }
+
+  // only need to calculate the out args
+  // for reduce type "max" and "min" for training
+  bool need_arg_out = at::GradMode::is_enabled()
+      && (input.requires_grad() || weight.requires_grad())
+      && (op == SPMM_MAX || op == SPMM_MIN);
+
+  if (!need_arg_out) {
+    spmm_reduce_stub(kCPU, out, crow, col, val, weight, op);
```

Comment:
Why not call normal `spmm` routine for REDUCE=SUM?

Reply:
actually, we can :) `spmm` routine will call into `mkl_sparse_spmm` and I choose not to use out of 3 reasons:

* mkl use `int32_t` as integer type and so it is very likely to require dtype conversion in pytorch where the indices are int64_t by default, which is a very heavy burden.
* mkl spmm did not cover `bfloat16` or `float16` which means i need 2 sets of kernels for different dtypes.
* mkl will use dynamic schedule in omp which is not quite suitable for the pyg scenario. The pyg scenario is more like: the payload is unbalanced by it is unbalanced in a certain manner. which means that we have a chance to pre-calculate a averaged payload and us static scheduling.

I actually have a set of test cases and benchmarks to test this hand-written kernel v.s. mkl spmm (to verify the correctness and speed), this one actually performs 2x better.

I have investigated into different implementation details before optimizing this one for pyg including MKL and the POC made by Intel Lab (they made optimization for DGL with hand written kernels, they also skip MKL).